### ISSUE TITLE
[HOLD for payment 2023-11-09] [$500] Web - New Dot Showing Console Error's when Creating New Policy from OldDot

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#28262]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open up OldDot and login with Email1.
2. Hover over settings > open policies.
3. Now in new tab open NewDot.
4. Login with Email2.
5. Now from Old Dot press "New Policy".
6. Verify that on navigating Console Error's are appearing and Also the other New Tab shows error with other Console Error's.

## Expected Result:
No Console Error Should be logged in either of the tabs of NewDot.

## Actual Result:
Console Error's are logged for both the tabs of NewDot.

## Workaround:
Unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.71.4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/720c8edd-1371-4d94-b134-feb0b2b9df08

https://github.com/Expensify/App/assets/93399543/d7e1eaca-6da7-4f20-bf00-5ddb5429d42f

**Expensify/Expensify Issue URL:**
**Issue reported by:** @jeet-dhandha 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1695429296097429

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c3a1c9a063a200e2</li>
        <li>Upwork Job ID: 1706660570183720960</li>
        <li>Last Price Increase: 2023-10-10</li>
<li>Automatic offers: </li>
<ul>
<li>jeet-dhandha | Contributor | 27300211</li>
<li>jeet-dhandha | Reporter | 27300215</li>
</ul></li>
    </ul>
</details>

**Comment by MrJithil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
While a different users signed in OldDot and NewDot, once a user clicks on the New Free Policy, a new tab will navigate the user to the NewDot with OldDot user login. It is creating console log errors. 

 
### What is the root cause of that problem?

Since the users are different, the api is returning empty `onyxData` for a couple of requests. This is not being passing through a null check. 

### What changes do you think we should make in order to solve the problem?


Currently we are checking the  `updateHandler(onyxData)`  for all api responses. Before each update, we are only checking the `onyxData` is `false` or not.

The `updateHandler` is accepting input parameter as an array. So, we only pass the `onyxData` to the `updateHandler` if it has at least an item. 

So, we need a check like `onyxData && onyxData.length ? call update handler : default resolver`

This change will be a single line change where we need to add the length check.
To comply with the code standards, we can make use of optional chain mechanism too. Ie,
`onyxData?.length ? call update handler : default resolver`



### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by kadiealexander** [comment]
@mollfpr please take a look!

**Comment by jeet-dhandha** [comment]
If you look at the error its trying to set an `SVG` in Onyx where only `string` type is allowed.

**Comment by jeet-dhandha** [comment]
Also the crash is being handled in another PR.

**Comment by jeet-dhandha** [comment]
But these two are different issues.

**Comment by mollfpr** [comment]
Could you attach the PR link?

**Comment by jeet-dhandha** [comment]
-  https://github.com/Expensify/App/issues/26530

**Comment by MrJithil** [comment]
> @MrJithil
> 
> > Since the users are different, the api is returning empty onyxData for a couple of requests. This is not being passing through a null check.
> 
> Which API request do you mean?
> 
> > Currently we are calling the updateHandler(onyxData) for all api responses. Before each update, we are only checking the onyxData is false or not.
> 
> What is the issue with the current implementation? Why it's showing the error in the opened tab and crash on existing tab?
> 
> @jeet-dhandha Why use `UserUtils.getDefaultAvatarURL` to solve this issue? How about the crash on the other tab?
> 
> > What is the root cause of that problem?
> > Below all places uses UserUtils.getDefaultAvatar function, which return's a SVG instead of default avatar URL.
> 
> Is this a root cause?

There are many api's. One of the example is OpenOldDotLink - `https://www.expensify.com/api?command=OpenOldDotLink` which is returning the empty `onyxData` . Screenshot attached. 
 
<img width="300" alt="Screenshot 2023-09-29 at 2 10 51 pm" src="https://github.com/Expensify/App/assets/26359740/fcc06ed1-0abc-49fe-a691-deaad1862a96">

Also, from the evidence I can see the error `Attempted to set
invalid data set in Onyx. Please ensure all data is
serializable. 15:57:00.667 [info] [Onyx] Failed`

<img width="200" alt="Screenshot 2023-09-29 at 3 01 35 pm" src="https://github.com/Expensify/App/assets/26359740/eec97142-f61f-4ce6-ad94-6dbcd778060a">

**Comment by mollfpr** [comment]
>No Console Error Should be logged in either of the tabs of NewDot.

@kadiealexander @jeet-dhandha It's not clear to me which log error we want to get rid; the expected was no log error in both tabs.

@MrJithil Could you show a video of the working result?

**Comment by MrJithil** [comment]
Sure. I can make a PR. I will find a way to point the old dot to local, record the video

**Comment by mollfpr** [comment]
>Sure. I can make a PR.

@MrJithil You don't have to. 

> will find a way to point the old dot to local

You can copy the link from the opened tab before it's loaded, and change the domain to your localhost.

**Comment by MrJithil** [comment]
> >Sure. I can make a PR.
> 
> 
> 
> @MrJithil You don't have to. 
> 
> 
> 
> > will find a way to point the old dot to local
> 
> 
> 
> You can copy the link from the opened tab before it's loaded, and change the domain to your localhost.

Okay. Sure. I will do that next day.

**Comment by MrJithil** [comment]
I can not able to replicate this issue anymore. Is there something merged in OldDot addressing this issues?

**Comment by mollfpr** [comment]
@MrJithil I still see the error log in the latest main. Which error can you replicate?

**Comment by MrJithil** [comment]
> @MrJithil I still see the error log in the latest main. Which error can you replicate?

Please attach the screenshot.

**Comment by mollfpr** [comment]
>It's not clear to me which log error we want to get rid; the expected was no log error in both tabs.

@kadiealexander @jeet-dhandha Could you guys help clarify it? Thanks!

**Comment by jeet-dhandha** [comment]
@mollfpr Update Proposal: https://github.com/Expensify/App/issues/28262#issuecomment-1735724144 
Changes:
- Added extra fix for new console error coming up for `openOnAdminRoom`.
- And added a Before and After Fix videos.

**Comment by jeet-dhandha** [comment]
> > It's not clear to me which log error we want to get rid; the expected was no log error in both tabs.
> 
> @kadiealexander @jeet-dhandha Could you guys help clarify it? Thanks!

We want to fix all error logs. But not the issue of `Empty Container`.

**Comment by sourcecodedeveloper** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Login to OD using account A. Login to ND using Account B. From OD open policies and click on New Policy. Open tab 1 where Account B is logged in and notice console error
TypeError: Cannot read properties of undefined (reading 'undefined')

Open tab 2 where the new policy tab is open. Notice the console error
Uncaught (in promise) DOMException: Failed to execute 'put' on 'IDBObjectStore'

The solution resolves console errors showing in both tabs.

### What is the root cause of that problem?

### What changes do you think we should make in order to solve the problem?


The console error in tab 1 can be resolve by placing null check 
Line 282 changes from
const shouldShowReportRecipientLocalTime = ReportUtils.canShowReportRecipientLocalTime(personalDetailsList, report, currentUserPersonalDetails.accountID) && !isComposerFullSize;

to

let shouldShowReportRecipientLocalTime = false;
if(personalDetailsList != undefined && personalDetailsList != null)
{
    shouldShowReportRecipientLocalTime = ReportUtils.canShowReportRecipientLocalTime(personalDetailsList, report, currentUserPersonalDetails.accountID) && !isComposerFullSize;
}

Also need to handle null check in OptionsRowLHNData.js

Line 185 changes from
key: ({fullReport}) => `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${fullReport.parentReportID}`,

to

 key: ({fullReport}) => 
 {
                 return  `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${fullReport != null ? fullReport.parentReportID : '0'}`;
 }


The console error in tab 2 Uncaught (in promise) DOMException: Failed to execute 'put' on 'IDBObjectStore' is occuring because the call to UserUtils.getDefaultAvatar(currentUserAccountID) returns function object. The function buildOptimisticCreatedReportAction() is then trying to store that object in IndexedDB but IndexedDB is unable to store function objects and therefore throws the exception. 
To continue storing function objects in IndexedDB, the property avatar needs to be non-serializable.
The function object is being stored in 3 functions
buildOptimisticClosedReportAction, buildOptimisticEditedTaskReportAction, buildOptimisticClosedReportAction() 

The avatar property can be non-serializable by calling Object.defineProperty and setting enumerable: false,

    Object.defineProperty(obj,"avatar",{
        value: lodashGet(allPersonalDetails, [currentUserAccountID, 'avatar'], UserUtils.getDefaultAvatar(currentUserAccountID)),
        enumerable: false
    });


          
### What alternative solutions did you explore? (Optional)

**Comment by sourcecodedeveloper** [comment]
@mollfpr [proposal ](https://github.com/Expensify/App/issues/28262#issuecomment-1752809919) attached.

**Comment by kadiealexander** [comment]
@mollfpr any updates here?

**Comment by mollfpr** [comment]
@jeet-dhandha Do you plan to fix the console error on the first tab?

@sourcecodedeveloper I don't know why we want to store the function object to the Onyx. All the avatar values are a URL.

**Comment by jeet-dhandha** [comment]
@mollfpr - if possible can you share the screenshot for first tab console error.

**Comment by sourcecodedeveloper** [comment]
> @jeet-dhandha Do you plan to fix the console error on the first tab?
> 
> @sourcecodedeveloper I don't know why we want to store the function object to the Onyx. All the avatar values are a URL.

@mollfpr The UserUtils.getDefaultAvatarUrl() that returns Url is only to support old dot avatar as it affects both platforms, for new dot the UserUtils.getDefaultAvatar() is to be used. The ideal approach would be to use UserUtils.getDefaultAvatar. The function object can still be used just needs to be marked as non-serializable and both the errors are resolved.

**Comment by sourcecodedeveloper** [comment]
> Screen.Recording.2023-10-10.at.23.55.10.mov
> 
> The PR you mentioned before is fixing the blank page on redirecting from OldDot.
> 
> The step to reproduce:
> 
>     1. Open the NewDot and sign in
> 
>     2. Open the OldDot on another tab and sign in with a different account
> 
>     3. Do the issue action perform
> 
> 
> You will see the first tab on **step 1** is broken, and there's a console error.
> 
> > We want to fix all error logs. But not the issue of Empty Container.
> 
> You mentioned that we want to fix all the error logs, so the crash on the first tab should be included, no?

@mollfpr The [proposed ](https://github.com/Expensify/App/issues/28262#issuecomment-1752809919) solution solves issues in both the tabs shown in that video, in case you missed it from the proposal.

**Comment by mollfpr** [comment]
@sourcecodedeveloper My concern is why we store the function object for the optimistic data but receive a string from the API response. Should the optimistic data align with the response from the API?

**Comment by sourcecodedeveloper** [comment]
@mollfpr I checked the api call for 'CreateWorkspace' when setting the optimistic data the avatar property is being set to function object i.e avatar : ƒ SvgDefaultAvatar15(props). The api response for 'CreateWorkspace' is not returning a string in fact it does not have the avatar property. 
I also checked api call for 'OpenReport' and the response from it does contain the avatar property and of type string like you said, 
"key": "personalDetailsList",
"avatar": "https:\/\/d2k5nsl2zxldvw.cloudfront.net\/images\/avatars\/default-avatar_24.png"
The explanation for the string response is that for the personalDetailsList old dot avatar is used meaning getDefaultAvatarURL is used and the api response would be string. So the only time the api response would be string is when the api needs to support old dot avatar.
If you can find an api call that is setting function object for avatar but the api response is string, please share as that would be an interesting scenario.

**Comment by jeet-dhandha** [comment]
First tab console logs should be fixed when the PR i mentioned fixes the blank window bug. 
- https://github.com/Expensify/App/issues/26530

For now we can focus on the newly opened tab only 👍 .

**Comment by kadiealexander** [comment]
@mollfpr any thoughts on the above?

**Comment by mollfpr** [comment]
Sorry for the delay; I just got out of the hospital for surgery on my right hand caused by motorcycle accident on Wednesday 🙏

![Screenshot 2023-10-16 at 12 31 31](https://github.com/Expensify/App/assets/25520267/ef627e6b-0cee-432e-8468-00807812a41f)

After reviewing the issue again, I agree with @jeet-dhandha that we can focus on the error log for the SVG function.

>The UserUtils.getDefaultAvatarUrl() that returns Url is only to support old dot avatar as it affects both platforms, for new dot the UserUtils.getDefaultAvatar() is to be used.

I disagree with this; the error appears when trying to save the optimistic data for some `reportActions_`, and @sourcecodedeveloper already proved that the SVG function couldn't be stored in the local storage. I don't think finding a way to store the function object would be ideal.

>I checked the api call for 'CreateWorkspace' when setting the optimistic data the avatar property is being set to function object i.e avatar : ƒ SvgDefaultAvatar15(props). The api response for 'CreateWorkspace' is not returning a string in fact it does not have the avatar property.

@sourcecodedeveloper You will see that the avatar is the string after refreshing the page and opening one of the reports. The response for the `reportActions_` avatar is a string.

In conclusion, the [proposal](https://github.com/Expensify/App/issues/28262#issuecomment-1735724144) from @jeet-dhandha has a better approach to solving the issue.

🎀 👀 🎀 C+ reviewed!

@kadiealexander Could you help again with the label? Thank you!

**Comment by sourcecodedeveloper** [comment]
@mollfpr the only time the avatar is the string is when getDefaultAvatarUrl() is being used, I could not find any example where avatar is being set to function object and the api response is string. If possible I would like to request to get second opinion from someone internal on replacing getDefaultAvatar() with getDefaultAvatarUrl(). Perhaps the internal engineer could at least better explain the use of function objects and if replacing getDefaultAvatar() with getDefaultAvatarUrl() is the correct approach.
@kadiealexander @mollfpr I request to please try to and add an internal engineer to the discussion and ask them for second opinion before rushing with the proposal. Thanks

**Comment by mollfpr** [comment]
@kadiealexander Friendly bump to re-adding the engineering label so @sourcecodedeveloper can prove their point. Thanks!

**Comment by sourcecodedeveloper** [comment]
@mollfpr Thanks.

**Comment by Li357** [comment]
It looks like it's a function because it's a React component that's generated by Webpack from importing an SVG. I agree with @jeet-dhandha's approach -- we shouldn't have React components in Onyx, so stopping them from being enumerated feels like a hack.

**Comment by mollfpr** [comment]
> [@mollfpr] The PR that introduced the bug has been identified. Link to the PR:

https://github.com/Expensify/App/pull/16581

> [@mollfpr] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

https://github.com/Expensify/App/pull/16581/files#r1386964381

> [@mollfpr] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

The regression step should be enough.

>[@mollfpr] Determine if we should create a regression test for this bug.
 [@mollfpr] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

**_The test case can be tested on the Web only._**


1. Open [www.expensify.com](https://www.expensify.com/) then login with user A.
2. Open [staging.new.expensify.com](https://staging.new.expensify.com/), then log in with user B.
3. Now go to Settings Tab > Workspaces.
4. Now, in Workspaces, go to the Group tab and press New Workspace on the top right.
5. Select Free workspace type.
6. As the new tab opens up, copy the existing URL and change its domain to staging.new.expensify.com
7. Now open inspect options and go to the Console's tab to verify no consoles are visible.
8. 👍 or 👎 

---

@kadiealexander, I think the PR is still eligible for bonus speed. The assignment is on 21st Oct, and the C+ approval is on 23rd Oct. Also, could you help with the payment summary for the manual request? Thank you!

**Comment by jeet-dhandha** [comment]
Any updates on payment ?

**Comment by mollfpr** [comment]
Friendly bump @kadiealexander

**Comment by kadiealexander** [comment]
Sorry team, the new K2 update has a "hide all held issues" checkbox that hid this!

**Comment by kadiealexander** [comment]
Payouts due:

- [ ] Issue Reporter: $50 @jeet-dhandha
- [ ] Contributor: $500+$250 bonus @jeet-dhandha
- [ ] Contributor+: $500+$250 bonus @mollfpr

Eligible for 50% #urgency bonus? [Yes](https://github.com/Expensify/App/pull/30115#issuecomment-1785546938)

Upwork job is [here](https://www.upwork.com/nx/wm/offer/27300211).

**Comment by JmillsExpensify** [comment]
$750 payment approved for @mollfpr based on summary above.
